InformationScripting/src/misc/Optional.h

```diff
@@ -0,0 +1,122 @@
+/***********************************************************************************************************************
+**
+** Copyright (c) 2011, 2015 ETH Zurich
+** All rights reserved.
+**
+** Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+** following conditions are met:
+**
+**    * Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+**      disclaimer.
+**    * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+**      following disclaimer in the documentation and/or other materials provided with the distribution.
+**    * Neither the name of the ETH Zurich nor the names of its contributors may be used to endorse or promote products
+**      derived from this software without specific prior written permission.
+**
+**
+** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**
+***********************************************************************************************************************/
+
+#pragma once
+
+#include "../informationscripting_api.h"
+
+namespace InformationScripting {
+
+/**
+ * Class inspired by std::experimental::optional, but it additionally contains an error/warning message.
+ */
+template <class ValueType>
+class INFORMATIONSCRIPTING_API Optional
+{
+		// Since the error constructor takes a QString we disallow usage with QString.
+		static_assert(!std::is_same<ValueType, QString>(), "QString is not allowed as paremeter for Optional");
+	public:
+		Optional(const ValueType& v, const QString& warning = {});
+		Optional(ValueType&& v, const QString& warning = {});
+		Optional(const QString& errorMessage);
+		Optional(QString&& errorMessage);
+
+		Optional(const Optional& other);
+		Optional(Optional&& other);
+
+		explicit operator bool() const;
+
+		const ValueType& value() const &;
+		ValueType& value() &;
+
+		ValueType&& value() &&;
+		const ValueType&& value() const &&;
+
+		bool hasError() const;
+		QString error() const;
+
+		bool hasWarning() const;
+		QString warning() const;
+	private:
+		enum class Type : int {Value, Warning, Error};
+		ValueType value_{};
+		QString message_;
```

Comment:
More abstractly, we need a way to describe why an optional is empty. In our particular case, this could be due to one or several errors, which were possibly preceded by one or several warnings. I agree it's somewhat heavy here, but on the bright side, the code where you create/handle the warnings/errors, doesn't change much.